control_plane/src/compute.rs

```diff
@@ -144,76 +144,60 @@ impl PostgresNode {
-            cfg_path.to_str().unwrap()
-        );
-        let port: u16 = CONF_PORT_RE
-            .captures(config.as_str())
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 1"))?
-            .iter()
-            .last()
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 2"))?
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 3"))?
-            .as_str()
+        let port: u16 = conf
+            .get("port")
+            .ok_or_else(|| anyhow!("could not find 'port' in config file {}", cfg_path_str))?
             .parse()
-            .with_context(|| err_msg)?;
+            .with_context(|| format!("could not parse 'port' in config_file {}", cfg_path_str))?;
 
         // parse timeline
-        let err_msg = format!(
-            "failed to find timeline definition in config file {}",
-            cfg_path.to_str().unwrap()
-        );
-        let timelineid: ZTimelineId = CONF_TIMELINE_RE
-            .captures(config.as_str())
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 1"))?
-            .iter()
-            .last()
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 2"))?
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 3"))?
-            .as_str()
+        let timelineid: ZTimelineId = conf
+            .get("zenith.zenith_timeline")
+            .ok_or_else(|| {
+                anyhow!(
+                    "could not find 'zenith.zenith_timeline' option in config file {}",
+                    cfg_path_str
+                )
+            })?
             .parse()
-            .with_context(|| err_msg)?;
+            .with_context(|| {
+                format!(
+                    "could not parse 'zenith.zenith_timeline' option in config file {}",
+                    cfg_path_str
+                )
+            })?;
 
         // parse tenant
-        let err_msg = format!(
-            "failed to find tenant definition in config file {}",
-            cfg_path.to_str().unwrap()
-        );
-        let tenantid = CONF_TENANT_RE
-            .captures(config.as_str())
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 1"))?
-            .iter()
-            .last()
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 2"))?
-            .ok_or_else(|| anyhow::Error::msg(err_msg.clone() + " 3"))?
-            .as_str()
+        let tenantid: ZTenantId = conf
+            .get("zenith.zenith_tenant")
+            .ok_or_else(|| {
+                anyhow!(
+                    "could not find 'zenith.zenith_tenant' option in config file {}",
+                    cfg_path_str
+                )
+            })?
             .parse()
-            .with_context(|| err_msg)?;
+            .with_context(|| {
+                format!(
+                    "could not parse 'zenith.zenith_tenant' option in config file {}",
+                    cfg_path_str
+                )
+            })?;
```

Comment:
There's a repeated pattern here of `.ok_or_else(...).parse().with_context(...)`, using approximately the same string.

This could be simplified a bit with a new method on `PostgresConf`:
```rust
impl PostgresConf {
    fn parse_field<T: FromStr>(&self, field_name: &str, file_name: &str) -> Result<T> {
        self
            .get(field_name)
            .ok_or_else(|| anyhow!("could not find '{}' option in config file {}", field_name, file_name))
            .parse()
            .with_context(|| format!("could not parse '{}' option in config file {}", field_name, file_name))
    }
}

// ----

let port: u16 = conf.parse_field("port", cfg_path_str)?;
let timelineid: ZTimelineId = conf.parse_field("zenith.zenith_timeline", cfg_path_str)?;
let tenantid: ZTenantId = conf.parse_field("zenith.zenith_tenant", cfg_path_str)?;
```

This solution isn't perfect (passing `cfg_path_str` everywhere isn't great), but I think it's probably an improvement. To make it better than what I'm suggesting, I'd probably add a separate `ConfFieldParser` type that pairs the config with the path and stores `cfg_path.to_string_lossy()` for you. IMO that's probably a bit overkill for just 3 fields.

Reply:
Ah, nice, changed it along those lines.